### PR TITLE
add fenv cache to task struct

### LINKED ISSUE #51277: add floating point environment to task state
I'd like to make the floating point environment (which controls rounding mode and floating point exceptions) a bit more useful. Currently we use the C behavior where the floating point environment is set per-thread, which doesn't really make sense in the context of tasks which can switch threads.

I'd like to propose adding the floating point environment (`fenv_t` object) to the task state. The simplest approach I can think of is that every time a thread switches tasks, we capture the current floating point environment of the current task and save it to the struct (via `fegetenv`), and set the environment to that of the new task (via `fesetenv`).  In this way, the floating point environment should follow the task (and be inherited by child tasks).

### PR DESCRIPTION
**Pull request by simonbyrne**
Fixes #51277. I know @vchuravy  is skeptical, but I wanted to try it out.

To give some examples:

```julia
julia> t = Base.Rounding.setrounding_raw(Float64, Base.Rounding.to_fenv(RoundDown)) do
           Task(() -> println(rounding(Float64)))
       end
Task (runnable) @0x000000010dff04c0

julia> rounding(Float64)
RoundingMode{:Nearest}()

julia> wait(schedule(t))
RoundingMode{:Down}() # currently gives RoundingMode{:Nearest}()

julia> rounding(Float64)
RoundingMode{:Nearest}()

julia> Base.Rounding.setrounding_raw(Float64, Base.Rounding.to_fenv(RoundDown)) do
           Threads.@threads :static for i = 1:Threads.nthreads()
               println(Threads.threadid() =>  rounding(Float64))
           end
       end
1 => RoundingMode{:Down}()
2 => RoundingMode{:Down}() # currently gives RoundingMode{:Nearest}()
4 => RoundingMode{:Down}() # currently gives RoundingMode{:Nearest}()
3 => RoundingMode{:Down}() # currently gives RoundingMode{:Nearest}()
```

### CODE REVIEW

File: src/task.c [hunk 1]
```diff
@@ -510,6 +510,8 @@ JL_NO_ASAN static void ctx_switch(jl_task_t *lastt)
     jl_set_pgcstack(&t->gcstack);
     jl_signal_fence();
     lastt->ptls = NULL;
+    fegetenv(&lastt->fenv);
```
Comment (vtjnash) [inline src/task.c:513]: This should come before the task switch, as some task switch implementations will clobber it

**Review by vtjnash — APPROVED** [review]
I think this should work

File: src/task.c (hunk 1 above)
Comment (simonbyrne) [reply src/task.c:513]: Where specifically should it go?

Comment (vtjnash) [reply src/task.c:513]: Or rather, it looks like the fesetenv is the one that needed to move

Comment (simonbyrne) [reply src/task.c:513]: Where should it go?

Comment (simonbyrne) [reply src/task.c:513]: Feel free to push to this branch

Comment (vtjnash) [reply src/task.c:513]: Already done. I don't know about the propagating behavior though. I think we might want to start every task with a call to set `fesetenv(FE_DFL_ENV)`?

**Review by vchuravy — DISMISSED** [review]
I am still not convinced that this is a good thing to do. We have not yet specified what `fenv` means on the language level and if we want to support it in these cases. Echoing https://github.com/JuliaLang/julia/issues/51277#issuecomment-1716941223

Doesn't longjmp/setjmp save fenv? E.g. why do we have to do manually?

**Comment by vtjnash** [comment]
I don't think so. That is a very significant problem for anyone using this API, but I don't think it means it is wrong to do this PR (e.g. https://github.com/lattera/glibc/blob/master/sysdeps/x86_64/jmpbuf-offsets.h and https://github.com/lattera/glibc/blob/master/sysdeps/x86_64/__longjmp.S have no space to be capable of storing it)

**Comment by vchuravy** [comment]
Two notes:

The C standard defines `fenv` as thread-local, I do agree that it would make sense to define it as a scoped value (as implemented here) and a task would copy the fence from its parent.

I do still maintain that doing so is preemptive without adding language level support for `strictfp` and [constrained floating point ops](https://llvm.org/docs/LangRef.html#constrained-floating-point-intrinsics) as defined by LLVM for this purpose.

Also see https://internals.rust-lang.org/t/equivalent-of-pragma-stdc-fenv-access-on/14934

I am okay with merging this partial support under the understanding that there is a declaration of intent to actually develop the full language semantics and that setting or accessing fenv remains UB.

Because right now messing with fenv and then entering the compiler that does some constant evaluation sounds like a recipe for causing hard to track down bugs.

**Comment by oscardssmith** [comment]
> might not be good to do, but it at least isn't worse than what we already do

I disagree with this. This has a real (if small) costs and unclear benefit.

**Comment by vtjnash** [comment]
Fair: I didn't expect AMD to seem so bad at this.
```
  CPU: 128 × AMD EPYC 7513 32-Core Processor
julia> @btime Base.Rounding.setrounding_raw(Float64, Base.Rounding.rounding_raw(Float64))
  19.679 ns (0 allocations: 0 bytes)

julia> t = @eval @task while true; yieldto($(current_task())); end
Task (runnable) @0x00007fae5e5b1090

julia> @btime yieldto(t)
  134.496 ns (0 allocations: 0 bytes)
```
```
  CPU: 10 × Apple M2 Pro
julia> @btime Base.Rounding.setrounding_raw(Float64, Base.Rounding.rounding_raw(Float64))
  5.916 ns (0 allocations: 0 bytes)
0

julia> @btime yieldto(t)
  67.197 ns (0 allocations: 0 bytes)
```

```
  CPU: 8 × Intel(R) Core(TM) i7-8809G CPU @ 3.10GHz
julia> @btime Base.Rounding.setrounding_raw(Float64, Base.Rounding.rounding_raw(Float64))
  6.380 ns (0 allocations: 0 bytes)
0

julia> @btime yieldto(t)
  175.516 ns (0 allocations: 0 bytes)
```

**Comment by JeffBezanson** [comment]
For me this more than doubled task switch time, from 120ns to 290ns. This at least needs to be discussed.

**Comment by DilumAluthge** [comment]
@JeffBezanson Does this still need to be discussed on a triage call? If so, I'll open a new issue. I don't think people notice closed PRs when filtering the triage label.

**Comment by DilumAluthge** [comment]
This PR was reverted in #56982, so I'll remove the triage label here.

**Pull request merged**